### DISCUSSION TITLE
None type object is not iterable for handling a trigger

### DISCUSSION THREAD

**Discussion by mnmneE** [wehr-lab/autopilot#119] [Q&A]
Hello,

while trying to simply toggle a LED by triggering a Digital In object(push button);
the pilot terminal throws an error about the PIN defined for the push button as it is not callable neither iterable;
I started from the blink script by adding ;

(
self.hardware['LEDS']['C'].set(1)
self.triggers['Pb'] = self.hardware['LEDS']['C'].toggle()
)

where Pb is defined in JSON as Digital_In object;
and LED C as Digital_Out object/
The error prevents me from assigning the  trigger to do any action ; after careful review of the tasks and how you use a Digital_In object to trigger a Digital_Out one either using the object self.target(for both POKES,LEDS and PORTS as the Free water task ) or by directly assigning the PIN name like the one I did (like self.triggers['Pb'] ;['C']or  ['E']);and after checking the JSON file, I am wondering what may cause the problem/as the problem refers to an error with the way the objects of the triggering event are defined;
Please find attached a screenshot of the error

in advance thanks very  for your help

Regards
![IMG-5937](https://user-images.githubusercontent.com/45367650/128501982-a51ea940-2b41-4ca0-b35d-f7245a087084.jpg)

**Comment by mnmneE** [comment]
I attach in this comment the JSON file that is used, as I don't know if I should specify the field of trigger ( of the  Digital_In object ) any instructions that may make the process run
![IMG-5938 (1)](https://user-images.githubusercontent.com/45367650/128530423-0b5378f3-5ecd-4a41-9b65-d2ea10a37bb5.jpg)

**Comment by sneakers-the-rat** [comment]
try 

```
(
self.hardware['LEDS']['C'].set(1)
self.triggers['Pb'] = self.hardware['LEDS']['C'].toggle
)
````

and let me know if that works?

note the lack of parentheses after toggle -- you're passing a reference to the method itself so that it can be called later, the parentheses call the method (which returns `None`)

**Reply by mnmneE** [reply]
Thanks for your answer , unfortunately I won’t have access till Monday , but just one remark, I tried also other methods , like set(‘1’) and it gave the same error,I will try and  I should have known it yet trying different methods led to the same result.

**Reply by sneakers-the-rat** [reply]
no problem! it's a subtle thing that I should emphasize more in the docs. The issue should be the same across other methods -- eg calling `set('1')` rather than passing `lambda: set('1')` as in https://github.com/wehr-lab/autopilot/blob/ced9dca4715f32324650fe4a0e0ecab8fd8965d8/autopilot/tasks/nafc.py#L274

which is not a common pattern in python at all, so I definitely think we need more elegant handling of triggers so that it's not so confusing. probably something like `set_trigger('Pb', self.hardware['group']['id'].method, '1')` for triggers that take arguments rather than needing to use a lambda function.

**Reply by sneakers-the-rat** [reply]
added to documentation requests https://github.com/wehr-lab/autopilot/issues/32

**Reply by mnmneE** [reply]
this is definitely the answer , thanks a lot
